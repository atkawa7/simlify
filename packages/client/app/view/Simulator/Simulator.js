import React from 'react';
import { connect } from 'react-redux';
import { alertActions, flowActions, nodeActions } from 'store/actions';
import NodeArea from 'components/NodeArea';
import SideComponents from 'components/SideComponents';
import SideHelper from 'components/SideHelper';
import TabBar from 'components/TabBar';

import ButtonBar from './ButtonBar';
import ImportDialog from './ImportDialog';
import validateFlowJSON from 'helper/validateFlowJSON';

import './Simulator.scss';

class Simulator extends React.Component {
  constructor() {
    super();

    this.state = {
      isImportDialogOpen: false,
    };

    this.nodeAreaRef = React.createRef();
  }

  UNSAFE_componentWillMount() {
    this.props.dispatch(nodeActions.loadAvailableNodes());
    this.props.dispatch(flowActions.loadFlow());
  }

  onTabChange(tabIndex) {
    this.props.dispatch(flowActions.changeCurrentFlowIndex(tabIndex));
  }

  handleSelectionChange(object, isSelected = false) {
    const selectedNodeModel = isSelected ? object : {};
    this.props.dispatch(nodeActions.setSelectedNodeModel(selectedNodeModel));
  }

  setImportDialogActive(isOpen) {
    this.setState({ isImportDialogOpen: isOpen });
  }

  importFlowData(jsonFlowString) {
    const jsonFlow = validateFlowJSON(jsonFlowString);
    if (!jsonFlow) {
      this.props.dispatch(alertActions.error('Not a valid Flow'));
    } else {
      this.props.dispatch(flowActions.createNewFlow(jsonFlow));
    }
  }

  onNodeAreaEvent(event) {
    if (event.function === 'selectionChanged') {
      this.handleSelectionChange(event.entity, event.isSelected);
    }
    if (event.function === 'entityRemoved') {
      this.handleSelectionChange(event.entity, false);
    }
  }

  render() {
    const { isImportDialogOpen } = this.state;
    const { flows, currentFlowIndex } = this.props.flowData;
    const { availableNodes, selectedNodeModel } = this.props.nodeData;

    const currentFlow = flows[currentFlowIndex];
    const tabs = flows
      .map(flow => ({ name: flow.name, icon: 'none' }));
    tabs.push({ name: 'Add', icon: 'none' });

    return (
      <div className="simulator">
        <ImportDialog
          open={isImportDialogOpen}
          onClose={() => this.setImportDialogActive(false)}
          onImport={(jsonData) => this.importFlowData(jsonData)}
        />
        <SideComponents
          availableNodes={availableNodes}
        />
        <div className="simulator__nodeArea">
          <ButtonBar 
            nodeAreaRef={this.nodeAreaRef}
            onImportClick={() => this.setImportDialogActive(true)}
          />
          <TabBar
            tabs={tabs}
            onTabChange={(selectedTab) => this.onTabChange(selectedTab)}
          />
          <NodeArea
            ref={this.nodeAreaRef}
            availableNodes={availableNodes}
            options={currentFlow}
            onEvent={(event) => this.onNodeAreaEvent(event)}
          />
        </div>
        <SideHelper
          nodeModel={selectedNodeModel}
        />
      </div>
    );
  }
};

const mapStateToProps = ({ flowData, nodeData }) => ({ flowData, nodeData });

export default connect(mapStateToProps)(Simulator);
