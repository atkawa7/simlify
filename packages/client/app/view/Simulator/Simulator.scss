@import "assets/scss/variables.scss";

.simulator {
    height: 100%;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
    display: flex;
    position: fixed;

    &__nodeArea {
        height: 100%;
        width: 100%;
        position: relative;
    }
}
