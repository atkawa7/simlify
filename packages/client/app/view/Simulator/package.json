{
    "name": "Simulator",
    "private": true,
    "main": "./Simulator.js"
}