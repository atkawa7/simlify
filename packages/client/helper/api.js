import axios from "axios";

const getRequest = async (url) => {
    const response = await axios.get(url);
    return response.data;
};

const postRequest = async (url, body) => {
    const response = await axios.post(url, body);
    return response.data;
};

const putRequest = async (url, body) => {
    const response = await axios.put(url, body);
    return response.data;
};

export default {
    getAvailableNodes: async () => await getRequest('http://localhost:3000/api/v1/nodes/list'),
    getFlows: async () => await getRequest('http://localhost:3000/api/v1/flows/all'),
    putFlow: async (updatedFlow) => await putRequest('http://localhost:3000/api/v1/flows', updatedFlow),
    postFlow: async (newFlow) => await postRequest('http://localhost:3000/api/v1/flows', newFlow),
};