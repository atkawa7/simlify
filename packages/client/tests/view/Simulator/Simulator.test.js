import React from 'react';
import api from '../../../helper/api';
import Simulator from '../../../app/view/Simulator';
import renderer from 'react-test-renderer';

jest.mock('../../../helper/api');

it('renders correctly', () => {
  api.getFlows.mockResolvedValue({});
  api.getAvailableNodes.mockResolvedValue({});

  const snapshot = renderer
    .create(<Simulator />)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});