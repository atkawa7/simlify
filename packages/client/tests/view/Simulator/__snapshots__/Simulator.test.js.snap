// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="simulator"
>
  <div
    className="simulator__nodeArea"
  >
    <div
      className="simulator__buttonBar"
    >
      <div
        className="simulator__buttonBar__container"
      >
        <div
          aria-label="small contained button group"
          className="MuiButtonGroup-root MuiButtonGroup-contained"
          role="group"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedContained MuiButton-contained MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Reset
            </span>
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedContained MuiButton-contained MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Save
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="tabBar"
    >
      <div
        className="tab tab--active"
        onClick={[Function]}
      >
        Add
      </div>
    </div>
    <div
      className="nodeArea"
      onDragOver={[Function]}
      onDrop={[Function]}
    >
      <div
        className="MuiPaper-root MuiPaper-elevation4 paper PaperComp-paper-32 nodeDragBar MuiPaper-rounded"
      />
      <div
        className="nodeArea__canvas css-18f0176"
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        onMouseUp={[Function]}
        onWheel={[Function]}
      >
        <svg
          className="css-ye4wrg"
          style={
            Object {
              "transform": "
			translate(
				0px,
				0px)
			scale(
				1
			)
  	",
            }
          }
        />
        <div
          className="css-ye4wrg"
          style={
            Object {
              "transform": "
			translate(
				0px,
				0px)
			scale(
				1
			)
  	",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="sideHelper"
  >
    <div
      className="tabBar"
    >
      <div
        className="tab tab--active"
        onClick={[Function]}
      >
        Description
      </div>
      <div
        className="tab "
        onClick={[Function]}
      >
        Settings
      </div>
    </div>
    <div
      className="sideHelper__body"
    />
  </div>
</div>
`;
